src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+            _documentVersionCache = documentVersionCache;
+            _documentMappingService = documentMappingService;
+        }
+
+        public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            int? documentVersion = null;
+            DocumentSnapshot documentSnapshot = null;
+            await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.RazorDocumentUri.GetAbsoluteOrUNCPath(), out documentSnapshot);
+
+                Debug.Assert(documentSnapshot != null, "Failed to get the document snapshot, could not map to document ranges.");
+
+                if (documentSnapshot is null ||
+                    !_documentVersionCache.TryGetDocumentVersion(documentSnapshot, out documentVersion))
+                {
+                    documentVersion = null;
+                }
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = null,
+                    HostDocumentVersion = null
+                };
+            }
+
+            var unmappedDiagnostics = request.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d)).ToArray();
+            if (!filteredDiagnostics.Any())
+            {
+                // No diagnostics left after filtering.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = Array.Empty<Diagnostic>(),
+                    HostDocumentVersion = documentVersion
+                };
+            }
+
+            return await MapDiagnosticsAsync(
+                request,
+                filteredDiagnostics,
+                documentVersion,
+                documentSnapshot).ConfigureAwait(false);
+
+            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
+            static bool CanDiagnosticBeFiltered(Diagnostic d) =>
+                (DiagnosticsToIgnore.Contains(d.Code?.String) &&
+                 d.Severity != DiagnosticSeverity.Error);
+        }
+
+        private async Task<RazorDiagnosticsResponse> MapDiagnosticsAsync(
+            RazorDiagnosticsParams request,
+            Diagnostic[] filteredDiagnostics,
+            int? documentVersion,
+            DocumentSnapshot documentSnapshot)
+        {
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
```

Comment:
I wouldn't make this assumption. I'd still map

Reply:
Updated. To confirm, this is to future proof for when `__virtual.html` pages don't map 1:1 with razor pages?

Reply:
Update; so just tried this out but doesn't seem like it's working. When we remove this guard, we end up utilizing the `DocumentMappingService.TryMapFromProjectedDocumentRange` to go from the "projected" range in the **HTML** document to the actual range. The `DocumentMappingService` is only designed for doing C# mapping ([works directly with the C# background document](https://github.com/dotnet/aspnetcore-tooling/blob/298c8ccd73e68d5221f0b5a20ade26eb74020d8d/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorDocumentMappingService.cs#L285)).

We're also making this assumption of skipping mapping for Razor/HTML [elsewhere](https://github.com/dotnet/aspnetcore-tooling/blob/298c8ccd73e68d5221f0b5a20ade26eb74020d8d/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageEndpoint.cs#L173). In which case, it may be worthwhile to keep this as is?

Not sure if I may have mis-understood the ask here, please do let me know if that's the case 😄

Reply:
> Updated. To confirm, this is to future proof for when `__virtual.html` pages don't map 1:1 with razor pages?

Correct



> The `DocumentMappingService` is only designed for doing C# mapping ([works directly with the C# background document](https://github.com/dotnet/aspnetcore-tooling/blob/298c8ccd73e68d5221f0b5a20ade26eb74020d8d/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorDocumentMappingService.cs#L285)).

Hmm, sounds like we may need to do more to decouple ourselves from there not being true HTML mappings. Can do that in another PR. Feel free to ignore the top level comment since it's a lot of work to properly fix

Reply:
Thanks for the clarification, created https://github.com/dotnet/aspnetcore/issues/29007.